### PR TITLE
BundlerEnv, now with groups and paths

### PR DESCRIPTION
**Pull request by nyarly**
This is a retry of
https://github.com/NixOS/nixpkgs/pull/23000

A crucial difference is that most of the machinery to assemble
a bundler working environment is now part of basic.nix,
and bundlerEnv uses that and then
binstubs in the specific gem it's being built for.

This is preliminary,
in that I'm thinking that there should also be
a builder expression to build a Gemfile
and then put links to explicit executables into $out.

At the moment though, I think this could be merged
as is:
it replaces all the functionality of bundlerEnv
and lays the groundwork for other things.

### CODE REVIEW

**Comment by nyarly** [comment]
One driver of a new PR was @zimbatm's comment that the existing bundlerEnv was already pretty complicated.

As I've been working through this again, I think that much of that complexity is inherent - or at least external to Nix. One of the core difficulties is that, if you're trying to install a gem-ified tool and it's specific requirements, there's kind of two ways in Rubyland to do that: the `gem install` way and the `bundle install` way. There could be a "Nix way" as well, but so far we've tried to hew as close to `bundle install` as possible, which is closer to the Nix approach.

A user of a Ruby gem, though, expects for binaries to be installed as described by the gemspec. The `buildRubyGem` builder parses a lot of that out into files when it builds a gem, but there's no way for a `bundlerEnv` expression to know _a priori_ what binaries a gem will produce, which reduces us to the bin-stubs.rb solution.

What I contemplate as a follow on is a rubyTool builder with explicit `binaries` and `scripts` attributes, which could then simply do a linkFarm of the binaries and wrappers of the scripts, again using `basic.nix` as a a basis. That'd also lead to easier-to-package Rails apps.

**Comment by nyarly** [comment]
Also in this PR: a ton of (still slightly messy) testing support. I found I wasn't able to keep all the concerns involved here straight between the occassional work sessions I've been able to make for it. After the fact I found `runTests` in debux.nix; that function is simpler, but the results are less complete. I'm biased, certainly.

**Comment by zimbatm** [comment]
thanks @nyarly, I will take a closer look this weekend

File: pkgs/development/ruby-modules/bundler-env/test.nix
Comment (zimbatm) [inline pkgs/development/ruby-modules/bundler-env/test.nix]: sweet!

File: pkgs/development/ruby-modules/bundler-env/test.nix
Comment (zimbatm) [inline pkgs/development/ruby-modules/bundler-env/test.nix]: awesome

File: pkgs/development/ruby-modules/bundler-env/tap-support.nix
Comment (zimbatm) [inline pkgs/development/ruby-modules/bundler-env/tap-support.nix]: I suppose that this is WIP?

File: pkgs/development/ruby-modules/bundler-env/functions.nix
Comment (zimbatm) [inline pkgs/development/ruby-modules/bundler-env/functions.nix]: Tell me if I misunderstood the purpose of this function.

When a gem is specified by path, all the files in the repo get copied into a derivation with the Gemfile. Now it would be a shame to make another copy of the gem into a new derivation. So instead this function is returned which pretends to be a derivation output but actually points into the other derivation. path would be something like `/nix/store/...-foobar/vendor/gems/mygem`.

File: pkgs/development/ruby-modules/bundler-env/functions.nix
Comment (zimbatm) [inline pkgs/development/ruby-modules/bundler-env/functions.nix]: for even better derivation simulation, add `__toString = self: self.outPath;`. That is a magic function that is used when coercing the attribute set into a string.

File: pkgs/development/ruby-modules/bundler-env/basic.nix
Comment (zimbatm) [inline pkgs/development/ruby-modules/bundler-env/basic.nix]: debug

File: pkgs/development/ruby-modules/bundler-env/basic.nix
Comment (zimbatm) [inline pkgs/development/ruby-modules/bundler-env/basic.nix]: it would be helpful to describe the responsibilities of the function since it's so complicated

**Review by zimbatm — COMMENTED** [review]
Really nice. I like that you even added a test framework.

I left a few comments and pushed a small fix commit, let me know when you have gone over them.

File: pkgs/development/ruby-modules/tool/default.nix
Comment (zimbatm) [inline pkgs/development/ruby-modules/tool/default.nix]: Is this file supposed to be in the top-level? I don't see any references to it yet.

**Review by zimbatm — COMMENTED** [review]
Had another round of review.

It seems like the "tool" module isn't linked in the top-level yet, is that the new entry-point for apps? Having a usage example would make the difference more apparent.

Also let me know if there is anything I can do to help.

**Comment by nyarly** [comment]
It is still a WIP. I mostly issued a PR as a heads up about the direction I'm going. With luck I'll have time to bring it all together this week, including a usage example.

**Comment by nyarly** [comment]
Okay - I think this is pretty much wrapped up. Would very much appreciate review, @zimbatm

File: pkgs/development/tools/corundum/default.nix
Comment (zimbatm) [inline pkgs/development/tools/corundum/default.nix]: is it possible to just set the `pname` and have the version derived from the gemset.nix?

File: pkgs/top-level/all-packages.nix
Comment (zimbatm) [inline pkgs/top-level/all-packages.nix]: maybe call it "bundlerApp" to keep the existing pattern?

**Review by zimbatm — COMMENTED** [review]
👍 overall. Just have a few comments inlined in the code.

File: pkgs/development/tools/corundum/default.nix
Comment (nyarly) [inline pkgs/development/tools/corundum/default.nix]: I think that's a really good idea. Doing it properly is going to take a dab more work - hold off merging this...

**Comment by Mic92** [comment]
It would be also nice to have an example in `nixpkgs/doc/language-frameworks/ruby.xml`.

**Comment by nyarly** [comment]
That's reasonable.

File: pkgs/development/ruby-modules/bundled-common/default.nix
Comment (0xABAB) [inline pkgs/development/ruby-modules/bundled-common/default.nix]: Useless quotes.

File: pkgs/development/ruby-modules/runtests.sh
Comment (0xABAB) [inline pkgs/development/ruby-modules/runtests.sh]: Just a single `pwd` output is impossible to find. Prefix it with something one can grep for or remove it.

File: pkgs/development/ruby-modules/testing/tap-support.nix
Comment (0xABAB) [inline pkgs/development/ruby-modules/testing/tap-support.nix]: Open an issue for this problem and reference the opened issue here.

**Review by 0xABAB — CHANGES_REQUESTED** [review]
Documentation for the Nix manual for this feature appears to be missing.

**Comment by nyarly** [comment]
I've added docs about bundlerApp - it seems presumptuous at this point to say "bundlerEnv is deprecated" - especially since there are several derivations that use it. Once this lands, perhaps a survey to see if bundlerEnv can be replaced by bundlerApp and if not, what shortcomings exist to be addressed.

File: doc/languages-frameworks/ruby.xml
Comment (0xABAB) [inline doc/languages-frameworks/ruby.xml]: the the

File: doc/languages-frameworks/ruby.xml
Comment (0xABAB) [inline doc/languages-frameworks/ruby.xml]: attributes, not items

**Review by 0xABAB — COMMENTED** [review]
Thanks for the documentation. This is nearing completion, I'd say.

File: doc/languages-frameworks/ruby.xml
Comment (0xABAB) [inline doc/languages-frameworks/ruby.xml]: unpredictable versions

**Comment by nyarly** [comment]
Anything else here?

**Comment by cstrahan** [comment]
@nyarly Thanks for the efforts!

I don't want to impede progress; however, I think the present state of the ruby/bundler stuff in Nixpkgs is plenty tricky (and I say that as the principal author), and I think some tweaks could be made to the PR to stave off further trickiness.

Here are the things that I find most excellent:

* A way to choose which executables to provide binstubs for (e.g. one can opt out of polluting one's $PATH with `rake` and such)
* A way to filter by the Gemfile groups
* Filtering by Ruby platform

Some of my concerns:

* There's a new `bundlerApp` abstraction, but I don't readily see why `bundlerEnv` itself couldn't be extended to support e.g. the `exes` argument (or any other new features).
* I find the new file and function indirection to be a little hard to follow. I know that's subjective, but I think a certain level of clarity is important to secure future contributions. I can provide more objective direction in this regard if desired.
* The TAP stuff could be the start of something great, but it appears that it's not ready, and is a little distracting at the moment. Probably best to factor out as a separate PR, no?
* As of now, this PR clocks in at +835 −131. That's a pretty sizable diff for what's added -- though you shouldn't take that as a slight against what this PR accomplishes, of course! I just think we can achieve a better cost:benefit ratio.

If we can address those few concerns, I'd love to get this merged in. I understand that, if we don't share the same aesthetic, it might be a bit bothersome to play "guess what would please @cstrahan", so if it's alright with you, I think I'll try to make the changes that I'd like and run them by you (hopefully over the next day or two). Alternatively, if you'd like to take a stab at this, I can make myself available here or on IRC.

**Comment by nyarly** [comment]
Hey @cstrahan - thanks for the feedback. I'm glad you're pleased by the main thrust here - constraining exes, selecting groups. Perhaps not obvious is that you're also able to use the (closer to) the full breadth of Bundler's Gemfile (e.g. the gemspec command) in nix-shell.

To your specific objections:
The division into multiple files was after consultation with @zimbatm, who had (effectively) the converse objection to the previous PR that did simply try to extend the `bundlerEnv` builder - much of the functionality was much more complicated as a result. Anyway, if there's a discussion about rolling this back to the original form (see #23000), I'd want to include him in that conversation.

Regarding the TAP stuff: I quickly found that the behaviors I was trying to add were more complicated that I could reasonably hand test. My Ruby-land instincts were to do as much TDD as I could manage (which I assume you can identify with), and queries in IRC came up with a kind of general shrug. I've since found that there's stuff in lib to do testing - if I'd known I wouldn't have written the TAP stuff! That said, I feel like there is value in the stuff I've started here, if only because the existing testing framework targets the integration level. With regards to it being half baked, I assume you're referring to the YAML stuff - it's an optional part of TAP which (if this framework were adopted more generally) it would be worth putting in the elbow grease to build YAML output etc. (If it isn't obvious, I do hope it is the start of something - testing in Nix was really pleasant.) 

Regarding TAP as a separate PR: maybe... but the motivation was the work in this PR, and on it's own I wonder if the utility of the testing framework would be evident. There's something of a chicken-and-egg problem there.

Regarding the size: while I agree that it's not a small PR, there's actually a fair amount going on. I realize in retrospect, I bit off an awful lot in starting this - although I contend there's not so many of us with the requisite expertise with Bundler, Rubygems _et al._ Better to have built a few more "package was missing or outdated" PRs first.

I appreciate your empathy about all this. This PR is months in the making, and does directly serve needs I have related to doing Ruby development work in Nix - one of the things that first drew me to Nix{,pkgs,OS} in the first. (I'm trying to contribute to RDF::Shex, which has moved on in the meantime...) So while there is the well-spotted "guess what will please @cstrahan" problem, there's also "please @zimbatm" and a handful of other contributors and the meta-game of "guess which contributor you need to please to get the PR merged" :) Maybe the thing to do is an RFC on the mailing list? (sadly you cannot see me flinch as I suggest that)

**Comment by nyarly** [comment]
@zimbatm @cstrahan I'd like to make progress on this, but at this point I'm at a loss about how to proceed.

It seems like, on the face of it, there's two potential approaches to adding the features that I need: extend the interface and function of the existing bundlerEnv (which is what #23000 did), or modularize into more, smaller components. My understanding over the last few months is that @cstrahan favors the former and @zimbatm the latter. 

Is there a way we can resolve this and get this merged?

**Comment by cstrahan** [comment]
> nyarly commented 17 days ago

Yikes, 17 days ago! With the way the last couple weeks have gone down, it feels more like half a week, at most. I feel really bad for leaving you hanging!

I'm reading through #23000 and also reviewing your feedback. Gimme a sec.

**Comment by cstrahan** [comment]
@nyarly 

> The division into multiple files was after consultation with @zimbatm, who had (effectively) the converse objection to the previous PR that did simply try to extend the bundlerEnv builder - much of the functionality was much more complicated as a result. Anyway, if there's a discussion about rolling this back to the original form (see #23000), I'd want to include him in that conversation.

Ah. I didn't realize that I was effectively pulling you in two directions.

> Regarding the TAP stuff: I quickly found that the behaviors I was trying to add were more complicated that I could reasonably hand test. My Ruby-land instincts were to do as much TDD as I could manage (which I assume you can identify with), and queries in IRC came up with a kind of general shrug.

Yeah, I totally appreciate that. I think that's an awesome initiative, and I definitely want to see some push-button approach to vetting that the whole thing works as expected.

> That said, I feel like there is value in the stuff I've started here, if only because the existing testing framework targets the integration level.

Definitely! My only concern was that if it wasn't _completely_ functional (and maybe it is -- feel free to correct me), it might add to the cognitive overhead of getting involved in this corner of Nixpkgs. I've been burned many times from jumping into a project where I crawl into some corner, take a look around and then start building up a mental model for the common vernacular, and the what-does-what and what-goes-where, only for someone (all to late!) to exclaim "oh, don't look there, that's _<dead-code / future-but-not-integrated-work / whatever>_!" So this quite possibly was a knee-jerk reaction, and it may not have been warranted (apologies!).

> I appreciate your empathy about all this. This PR is months in the making, and does directly serve needs I have related to doing Ruby development work in Nix - one of the things that first drew me to Nix{,pkgs,OS} in the first.

I know how it can be. Getting the initial bits of the Bundler support in place was incredibly taxing (my [PR](https://github.com/rubygems/rubygems.org/pull/713) to add SHAs to Rubygems.org languished due to them deciding to rewrite the git history right around the same time, and suggestions to Bundler to make things more amenable to packaging fell on deaf ears). I presume you're getting a taste of that here, and I apologize for that. I hope (and I suspect) that things will go little more smoothly in future endeavors. Sadly, interest in Ruby is a little underrepresented in the Nix community (though certainly non-zero), which I think is another chicken-and-egg thing: if the Ruby stuff is painful, why would any Rubyists want to invest in improving any of it? I hope things will improve and that things will no longer be blocked on just me and/or @zimbatm, and as such, your interest - and more importantly, involvement - is very, very much appreciated!

> Maybe the thing to do is an RFC on the mailing list? (sadly you cannot see me flinch as I suggest that)

A confession: I haven't read a single personal email in about 2 months. Sounds crazy, I know, but I figure if it's important, someone will either ring me or meet me in meat-space. I've been really strung out and oversubscribed, and something - for the sake of my sanity - had to give. Email - what with the incessant recruiter spam, shitty product ads, Yelp/Amazon/dozens-of-similar-ilk alerts that consistently evade my filters (they keep changing the domain names and such; none of them obey my request to unsubscribe; and I can't mark them as spam or I'll risk _not_ receiving the shit I _do_ want, like order receipts) - was just the thing I needed a break from. Thus, I'm now a proud "inbox 3K" person, and it feels great. I'll probably have to get back to reality at some point, but for now I'm going to savor it.

In short, I had no idea you emailed the mailing list :stuck_out_tongue_winking_eye:.

----

Alright, so about the fate of this PR. I _really_ want all this stuff in Nixpkgs - the _only_ thing that I have the slightest concern with is the _shape_ of the thing. Function-wise, it's solid :+1:.

So here's where I feel bad -- I actually really like the idea of extending the underlying `bundlerEnv` directly, rather than adding a separate thing. From looking at the [last comment on #23000](https://github.com/NixOS/nixpkgs/pull/23000#issuecomment-292819289), it looks like you discussed this with @zimbatm somewhere else (perhaps IRC); maybe if I knew what sort of [pros and cons](https://www.reddit.com/r/SiliconValleyHBO/comments/36btml/the_swot_board_about_blaine/) were sorted out, I might change my stance. Right now, I _think_ we can keep the convenience of one go-to solution for Nix-ifying a Bundler setup, without adding too much complexity to what's already there. But I also appreciate your position of being pulled between two reviewers.

**So here's my thinking (i.e. TL;DR):** instead of pulling you in two directions, lets accept this **now**. The onus will then be on me to submit any subsequent PRs for your and @zimbatm's review (and any other interested parties, should they exist). We have about a month until the 17.09 branch-off, so I'll implement and propose any breaking changes within that window. That should get your changes in and avoid bit-rot, and hopefully feel a lot more fair than the current state of things. The one bit of possible difficulty on your part will (potentially) be upkeep with any changes that make it in between now and the branch off; it's my hope, though, that this makes for a decent compromise on the whole.

How does that sound? If that's all good, I'll go ahead and pull the trigger.

**Comment by nyarly** [comment]
@cstrahan You have my sympathies - I know how things can get to be that way (i.e. avoiding email etc), and it usually sucks.

**I'm 100% behind merging this and working forward on it.** I look forward to seeing the next iteration.

I would like to have the conversation about expanding bundlerEnv vs. splitting it based on use case etc. Something to find time with @zimbatm on, if we can.

Separately, I'd love to expand the question of testing - totally agree that the TAP stuff should either conquer or die, and the "oh, not that code" is exactly the guiding principle. If it dies, I would really like there to be a replacement though. (I do think that the testing layer itself is reasonably functional - there's some shell scripts to drive it, but they're just shims around nix-build). Who can we bring in on that?

**Comment by cstrahan** [comment]
> I'm 100% behind merging this and working forward on it. I look forward to seeing the next iteration.

Sweet!

> I would like to have the conversation about expanding bundlerEnv vs. splitting it based on use case etc. Something to find time with @zimbatm on, if we can.

Sounds good. I'll try to make myself available over IRC over the next week or so (I get push-notifications to my phone, so that's actually a pretty reliable way to get a hold of me, assuming I'm not asleep when the notification arrives).

> Separately, I'd love to expand the question of testing - totally agree that the TAP stuff should either conquer or die, and the "oh, not that code" is exactly the guiding principle. If it dies, I would really like there to be a replacement though. (I do think that the testing layer itself is reasonably functional - there's some shell scripts to drive it, but they're just shims around nix-build). Who can we bring in on that?

Not sure about who to bring in. @pikajude has done a bunch of great stuff Ruby-wise, but I haven't seen him in a while, so I don't know how active he is / wants to be. There's the chance that I can help out, but... we both saw how well that worked out here :wink:. Other people might be more responsive on the mailing list, so that might be a good way to drum up support.

**Comment by NeQuissimus** [comment]
This looks good to me, I wish I'd had this 6 months ago when I was doing Ruby :D

**Comment by joachifm** [comment]
I think the manual stopped building after this, see https://travis-ci.org/NixOS/nixpkgs/jobs/258787424

**Comment by fpletz** [comment]
nixpkgs manual build fixed: 4d66de88fffada1f4970e5173be78c1ef30a8798

**Comment by joachifm** [comment]
@fpletz thanks :)

**Comment by joachifm** [comment]
Looks like there might be some problems still: https://travis-ci.org/NixOS/nixpkgs/jobs/258903666

**Comment by globin** [comment]
Fixed up further in 0154d1b03fc, could you check if I interpreted everything correctly?

**Comment by zimbatm** [comment]
woot! thanks @nyarly @cstrahan!

**Pull request merged**
